### PR TITLE
Add set and get thread name prctl options

### PR DESCRIPTION
**Pull request by vtikoo**
Signed-off-by: Vikas Tikoo <vitikoo@microsoft.com>

### CODE REVIEW

File: include/myst/thread.h [hunk 1]
```diff
@@ -187,6 +187,9 @@ struct myst_thread
     myst_spinlock_t* thread_lock;
     struct myst_thread* group_prev;
     struct myst_thread* group_next;
+
+    /* thread name */
+    char name[16];
```
Comment (jxyang) [inline include/myst/thread.h:192]: Is this what Linux expected?

File: kernel/thread.c [hunk 2]
```diff
@@ -680,6 +687,13 @@ static long _syscall_clone_vfork(
         child->run_thread = myst_run_thread;
         child->main.thread_group_lock = MYST_SPINLOCK_INITIALIZER;
         child->thread_lock = &child->main.thread_group_lock;
+        {
+            char buf[BUFSIZ];
+            /* ATTN: we don't take a lock on _num_threads,
+             thread names could be duplicates */
+            snprintf(buf, sizeof(buf), "thread-%ld", _num_threads);
+            myst_set_thread_name(child, buf);
```
Comment (jxyang) [inline kernel/thread.c:695]: Why don't we just sprintf into the myst_thread_t name buffer?

File: kernel/thread.c [hunk 3]
```diff
@@ -605,7 +605,14 @@ static long _syscall_clone(
         child->crt_td = newtls;
         child->run_thread = myst_run_thread;
         child->thread_lock = parent->thread_lock;
-
+        {
+            char buf[BUFSIZ];
+            /* ATTN: we don't take a lock on _num_threads,
+             thread names could be duplicates */
+            snprintf(buf, sizeof(buf), "thread-%ld", _num_threads);
+            myst_set_thread_name(child, buf);
```
Comment (jxyang) [inline kernel/thread.c:613]: Same as below.

File: kernel/thread.c [hunk 4]
```diff
@@ -813,3 +827,23 @@ size_t myst_kill_thread_group()
...
+{
+    int ret = 0;
+
+    if (!thread || !n)
+        ERAISE(-EINVAL);
+
+    /* Copy atmost 15 bytes */
+    strncpy(thread->name, n, sizeof(thread->name) - 1);
+    /* Set null terminator if string pointed by n
+        longer than the 16-byte fixed buffer for thread->name */
```
Comment (jxyang) [inline kernel/thread.c:841]: Is this the same behavior as Linux?

File: include/myst/thread.h (hunk 1 above)
Comment (vtikoo) [reply include/myst/thread.h:192]: Yes. 

https://elixir.bootlin.com/linux/latest/source/include/linux/sched.h#L957
https://elixir.bootlin.com/linux/latest/source/include/linux/sched.h#L215

**Comment by jxyang** [comment]
Can you check if the document on syscall limitations should be updated?

File: kernel/thread.c (hunk 4 above)
Comment (vtikoo) [reply kernel/thread.c:841]: Yes. According to the man prctl page - the name can be atmost 16 bytes, including the null termination character.

File: kernel/thread.c (hunk 2 above)
Comment (vtikoo) [reply kernel/thread.c:695]: Probably a theoritical concern, but `_num_threads` being a unsigned long int, could cause the thread name to go over 16 bytes. So this was delegating to `myst_set_thread_name`, which does the bounds checking anyways.

Comment (jxyang) [reply kernel/thread.c:695]: We can check `_num_threads` here. Checking an integer is always faster than checking a string.

Comment (vtikoo) [reply kernel/thread.c]: Changed to not use `myst_set_thread_name`.

File: kernel/thread.c [hunk 5]
```diff
@@ -680,6 +687,13 @@ static long _syscall_clone_vfork(
         child->run_thread = myst_run_thread;
         child->main.thread_group_lock = MYST_SPINLOCK_INITIALIZER;
         child->thread_lock = &child->main.thread_group_lock;
+        {
+            char buf[16];
+            /* ATTN: we don't take a lock on _num_threads,
+             thread names could be duplicates */
+            snprintf(buf, sizeof(buf), "thread-%ld", _num_threads % 99999999);
```
Comment (jxyang) [inline kernel/thread.c:694]: Actually, I meant to replace the whole block with one line:
```
snprintf(child->name, sizeof(buf), "thread-%ld", _num_threads % 99999999);
```

Comment (vtikoo) [reply kernel/thread.c:694]: Oh right, thats much simpler 👍

Comment (vtikoo) [reply kernel/thread.c]: Fixed.

**Review by jxyang — APPROVED** [review]

**Pull request merged**